File: mobile/lib/core/api/galat_api.dart
/// Kegagalan saat bicara dengan backend.
///
/// Dibedakan menurut apa yang harus dilakukan pengguna, bukan menurut kode HTTP-nya.
/// Layar tidak perlu tahu angka 401 atau 409; yang perlu ia tahu cuma "ini bisa
/// dicoba lagi" atau "ini butuh tindakan lain". Menyebar `if (status == 401)` ke
/// banyak layar membuat setiap layar harus ingat aturan yang sama, dan cepat atau
/// lambat ada satu yang lupa.
sealed class GalatApi implements Exception {
  const GalatApi(this.pesan);

  /// Kalimat yang layak ditampilkan apa adanya ke pengguna.
  final String pesan;

  @override
  String toString() => '$runtimeType: $pesan';
}

/// Tidak sampai ke server sama sekali: tidak ada sinyal, server mati, atau timeout.
///
/// Dipisahkan dari galat server karena tindakannya berbeda. Ini satu-satunya jenis
/// yang pantas ditawari tombol "coba lagi", sebab mencoba ulang memang bisa berhasil
/// tanpa ada yang berubah.
class GalatJaringan extends GalatApi {
  const GalatJaringan([super.pesan = 'Tidak bisa menghubungi server. Periksa koneksimu.']);
}

/// Permintaan ditolak karena isinya, misal nomor HP yang bentuknya salah.
class GalatPermintaan extends GalatApi {
  const GalatPermintaan(super.pesan);
}

/// Belum masuk, atau kredensialnya tidak cocok.
///
/// Untuk permintaan yang membawa token, ini berarti tokennya sudah kedaluwarsa atau
/// dicabut, dan pengguna harus masuk ulang.
class GalatTidakBerwenang extends GalatApi {
  const GalatTidakBerwenang([super.pesan = 'Sesi kamu sudah berakhir. Masuk lagi, ya.']);
}

/// Sudah masuk, tapi memang tidak berhak melakukannya.
class GalatDilarang extends GalatApi {
  const GalatDilarang([super.pesan = 'Kamu tidak punya akses untuk melakukan ini.']);
}

/// Yang dicari tidak ada, atau ada tapi bukan milik pemanggil.
///
/// Keduanya sengaja tidak dibedakan, mengikuti backend yang menjawab 404 untuk
/// order milik orang lain. Membedakannya di sini akan membocorkan lagi apa yang
/// sudah susah payah ditutup di sana.
class GalatTidakDitemukan extends GalatApi {
  const GalatTidakDitemukan([super.pesan = 'Data yang dicari tidak ada.']);
}

/// Bentrok dengan keadaan yang sudah ada, misal nomor HP yang sudah terdaftar.
class GalatBentrok extends GalatApi {
  const GalatBentrok(super.pesan);
}

/// Ditolak karena terlalu sering, bukan karena isinya salah.
///
/// Dipisahkan dari [GalatServer] justru karena tindakannya berlawanan. Untuk galat
/// server, "coba lagi sebentar lagi" adalah saran yang benar; di sini mencoba lagi
/// adalah persis hal yang membuatnya ditolak. Yang harus disampaikan ke pengguna
/// adalah menunggu, dan berapa lama.
///
/// Pesannya datang dari server, bukan dikarang di sini, karena hanya server yang tahu
/// batas mana yang tercapai dan berapa lama sisanya.
class GalatTerlaluSering extends GalatApi {
  const GalatTerlaluSering([
    super.pesan = 'Terlalu sering mencoba. Tunggu sebentar, lalu coba lagi.',
  ]);
}

/// Server yang bermasalah, bukan permintaannya.
///
/// Pesannya sengaja tidak memuat isi jawaban server. Jejak galat backend yang
/// sampai ke layar pengguna adalah bocoran gratis tentang bentuk dalam sistem,
/// dan tidak menolong siapa pun yang sedang memesan ojek.
class GalatServer extends GalatApi {
  const GalatServer([super.pesan = 'Server sedang bermasalah. Coba lagi sebentar lagi.']);
}

File: web/src/inti/galat_api.ts
/**
 * Kegagalan saat bicara dengan backend.
 *
 * Kembaran `mobile/lib/core/api/galat_api.dart`, dan pembagiannya sengaja sama:
 * menurut apa yang harus dilakukan orang yang menerimanya, bukan menurut kode HTTP-nya.
 * Layar tidak perlu tahu angka 401 atau 409; yang perlu ia tahu cuma "ini bisa dicoba
 * lagi", "ini harus masuk ulang", atau "ini memang tidak boleh".
 *
 * Sebabnya bukan kerapian. Menyebar `if (status === 401)` ke banyak layar berarti setiap
 * layar harus ingat aturan yang sama, dan cepat atau lambat ada satu yang lupa. Di
 * dashboard admin taruhannya lebih besar daripada di aplikasi mobile: satu layar yang
 * salah membaca 403 sebagai "server error" akan menyuruh admin mencoba lagi berkali-kali
 * untuk hal yang memang tidak akan pernah diizinkan.
 */
export abstract class GalatApi extends Error {
  // Publik walaupun kelasnya abstrak: subkelas seperti GalatPermintaan sengaja tidak
  // menulis ulang konstruktornya, dan konstruktor terlindung membuat subkelas itu ikut
  // tidak bisa dibuat dari luar berkas ini.
  constructor(pesan: string) {
    super(pesan);
    this.name = new.target.name;
  }
}

/** Tidak sampai ke server sama sekali: server mati, jaringan putus, atau kelewat batas waktu. */
export class GalatJaringan extends GalatApi {
  constructor(pesan = 'Tidak bisa menghubungi server. Periksa koneksi dan pastikan backend menyala.') {
    super(pesan);
  }
}

/** Permintaan ditolak karena isinya, misal kata kunci pencarian yang terlalu pendek. */
export class GalatPermintaan extends GalatApi {}

/**
 * Belum masuk, atau tokennya sudah kedaluwarsa.
 *
 * Satu-satunya jenis yang boleh mengeluarkan orang dari dashboard, dan itu ditangani di
 * satu tempat (lihat penyedia sesi), bukan di setiap layar.
 */
export class GalatTidakBerwenang extends GalatApi {
  constructor(pesan = 'Sesi kamu sudah berakhir. Masuk lagi, ya.') {
    super(pesan);
  }
}

/**
 * Sudah masuk, tapi memang tidak berhak.
 *
 * Di dashboard ini artinya hampir selalu satu hal: akunnya tidak menyandang peran admin.
 * Dibedakan dari [GalatTidakBerwenang] justru supaya orangnya tidak dilempar ke layar
 * masuk berulang-ulang untuk masalah yang tidak akan selesai dengan masuk ulang.
 */
export class GalatDilarang extends GalatApi {
  constructor(pesan = 'Akun ini tidak punya peran admin.') {
    super(pesan);
  }
}

/** Yang dicari tidak ada. */
export class GalatTidakDitemukan extends GalatApi {
  constructor(pesan = 'Data yang dicari tidak ada.') {
    super(pesan);
  }
}

/** Bentrok dengan keadaan yang sudah ada. */
export class GalatBentrok extends GalatApi {}

/**
 * Ditolak karena terlalu sering, bukan karena isinya salah.
 *
 * Pesannya datang dari server, bukan dikarang di sini, karena cuma server yang tahu batas
 * mana yang tercapai dan berapa lama sisanya.
 */
export class GalatTerlaluSering extends GalatApi {
  constructor(pesan = 'Terlalu sering mencoba. Tunggu sebentar, lalu coba lagi.') {
    super(pesan);
  }
}

/**
 * Server yang bermasalah, bukan permintaannya.
 *
 * Pesannya sengaja tidak memuat isi jawaban server. Jejak galat backend yang sampai ke
 * layar adalah bocoran gratis tentang bentuk dalam sistem.
 */
export class GalatServer extends GalatApi {
  constructor(pesan = 'Server sedang bermasalah. Coba lagi sebentar lagi.') {
    super(pesan);
  }
}

/**
 * Memilih jenis galat dari kode status, dan mengambil kalimatnya dari ProblemDetails
 * kalau ada.
 *
 * Backend menjawab hampir seluruh penolakannya sebagai ProblemDetails (`AddProblemDetails`
 * di Program.cs), dan kalimat di dalamnya ditulis untuk dibaca manusia. Membuang kalimat
 * itu lalu menggantinya dengan karangan sendiri berarti admin membaca "permintaan ditolak"
 * padahal server sudah menyebutkan persis apa yang kurang.
 *
 * Kecualinya 5xx: kalimat dari sana bukan untuk pengguna.
 */
export function galatDari(status: number, badan: unknown): GalatApi {
  const detail = pesanProblemDetails(badan);

  switch (true) {
    case status === 400 || status === 422:
      return new GalatPermintaan(detail ?? 'Permintaan ditolak server.');
    case status === 401:
      return new GalatTidakBerwenang();
    case status === 403:
      return new GalatDilarang();
    case status === 404:
      return new GalatTidakDitemukan();
    case status === 409:
      return new GalatBentrok(detail ?? 'Keadaannya sudah berubah. Muat ulang halamannya.');
    case status === 429:
      return detail ? new GalatTerlaluSering(detail) : new GalatTerlaluSering();
    default:
      return new GalatServer();
  }
}

function pesanProblemDetails(badan: unknown): string | null {
  if (typeof badan !== 'object' || badan === null) return null;

  const isi = badan as { title?: unknown; detail?: unknown };
  const judul = typeof isi.title === 'string' ? isi.title.trim() : '';
  const rincian = typeof isi.detail === 'string' ? isi.detail.trim() : '';

  // Keduanya digabung karena backend memang memakai keduanya bersamaan: judul menyebut
  // apa yang ditolak, detail menyebut apa yang harus dilakukan. Menampilkan judulnya saja
  // membuang separuh yang berguna ("Ini admin terakhir" tanpa "angkat admin lain dulu").
  const gabungan = [judul, rincian].filter(Boolean).join('. ');
  return gabungan.length > 0 ? gabungan : null;
}
